File: Proc-0/src/main/kotlin/Interop/FileWriter.kt
package proc.Interop

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import kotlin.math.log2

/**
 * Файловый поток для записи.
 */
class FileWriter(fileName: String): Writer {

	private var _handle: File
	private var _outputStream: FileOutputStream

	init {
		_handle = File(fileName)
		_outputStream = _handle.outputStream()
	}

	companion object {

		/**
		 * Compiler is not support state logic.
		 * TODO Эту фигню надо будет заменить таблицей.
		 */
		fun BaseToBytes(baseObj: Int): Int {
			return log2(baseObj as Double) as Int / 8
		}
	}

	fun Close() {
		_outputStream.close()
	}

	override fun Write(source: Reader): Any {
		TODO("Not yet implemented")
	}

	override fun Write(source: Any, baseObj: Any): Any {
		var byteCount: Int = BaseToBytes(baseObj as Int)
		_outputStream.write(source as ByteArray, 0, byteCount)
		return byteCount
	}
}

File: Proc-0/Interop/FileWriter.cs
/*
	2. Универсальная машина не может работать без входных и выходных данных. Мне еще не известно, как эти данные должны
	выглядеть и передаваться. Поэтому необходимо ввести некоторую абстракцию для работы с потоками, известными как
	ленты. Каждая клетка ленты может содержать несколько различных значений, как в машине Тьюринга. На отдельных
	архитектурах такие клетки могут быть основаны на разных физических принципах и могут кодировать количество
	состояний - оснований, а объём данных можно выразить через логарифм общего количества состояний по основанию
	архитектуры. Такое представление объёма, к сожалению, возрастает экспоненциально, но полностью отвязывает код от
	оборудования.
*/

namespace Interop;

// Если использовать interface segregation, то может и не будет смысла в объединённом интерфейсе.
// Также есть идея объединений для безымянных композиций интерфейсов и др. типов.
//class Stream: public Reader, public Writer, public Seeker {
//};

/// <summary>
/// Файловый поток.
/// </summary>
class FileWriter : Writer
{
	private System.IO.FileStream _handle;

	public FileWriter(string fileName, FileMode fileMode, FileAccess fileAccess)
	{
		Open(fileName, fileMode, fileAccess);
		if (_handle == null)
			throw new Exception("Can't create file stream.");
	}

	/// <summary>
	/// Compiler is not support state logic.
	/// TODO Эту фигню надо будет заменить таблицей.
	/// </summary>
	public static int BaseToBytes(int baseobj)
	{
		return (int)Math.Log2(baseobj) / 8;
	}

	public bool Open(string fileName, FileMode fileMode, FileAccess fileAccess)
	{
		_handle = new System.IO.FileStream(fileName, fileMode, fileAccess);
		return true;
	}

	public void Close()
	{
		_handle.Close();
	}

	public object Write(object source, object baseobj)
	{
		int byteCount = BaseToBytes((int)baseobj);
		_handle.Write((byte[])source, 0, byteCount);
		return byteCount;
	}

	public object Write(Reader source)
	{
		throw new NotImplementedException();
	}
}
